File: backend/app/models/yamnet.py
import tensorflow as tf
import tensorflow_hub as hub
import numpy as np
import os

class SoundClassifier:
    def __init__(self):
        # Load YAMNet model from TF Hub
        # In a real production app, you might download this locally first
        print("Loading YAMNet model...")
        self.model = hub.load('https://tfhub.dev/google/yamnet/1')
        self.class_map_path = self.model.class_map_path().numpy()
        self.class_names = self._load_class_names(self.class_map_path)
        print("Model loaded successfully.")

    def _load_class_names(self, class_map_path):
        import csv
        class_names = []
        with tf.io.gfile.GFile(class_map_path) as csvfile:
            reader = csv.DictReader(csvfile)
            for row in reader:
                class_names.append(row['display_name'])
        return class_names

    def classify(self, audio_np):
        """
        Classifies audio and returns the most specific top prediction.
        audio_np: 1-D float32 numpy array of 16kHz audio.
        """
        scores, embeddings, spectrogram = self.model(audio_np)

        # Average scores over all frames
        mean_scores = np.mean(scores, axis=0)
        top5_indices = np.argsort(mean_scores)[-5:][::-1]

        # Generic labels that obscure specific sounds — skip them
        GENERIC = {
            'Silence', 'Background noise', 'Noise', 'Sound', 'Environmental noise',
            'Inside, small room', 'Inside, large room or hall',
            'Outside, urban or manmade', 'Outside, rural or natural',
            'Music', 'Speech', 'Narration, monologue',
        }

        # Find the first (highest-scored) non-generic label
        best_idx = top5_indices[0]
        for idx in top5_indices:
            name = self.class_names[idx]
            if name not in GENERIC:
                best_idx = idx
                break

        label = self.class_names[best_idx]
        confidence = float(mean_scores[best_idx])

        return {
            "label": label,
            "confidence": confidence,
            "all_predictions": [
                {"label": self.class_names[i], "confidence": float(mean_scores[i])}
                for i in top5_indices
            ]
        }

# Singleton instance
classifier = None

def get_classifier():
    global classifier
    if classifier is None:
        classifier = SoundClassifier()
    return classifier

File: frontend/src/utils/yamnet.js
import * as tf from '@tensorflow/tfjs';
import { YAMNET_LABELS } from './yamnetLabels';

// YAMNet as a TF.js graph model. The base TFHub path resolves to the model
// files when `fromTFHub` is set; the GCS path is a direct mirror.
const MODEL_URLS = [
  { url: 'https://tfhub.dev/google/tfjs-model/yamnet/tfjs-graph-model/1', fromTFHub: true },
  { url: 'https://storage.googleapis.com/tfjs-models/savedmodel/yamnet/model.json', fromTFHub: false },
];

// Generic labels to skip — return the more specific sound instead
const GENERIC = new Set([
  'Silence', 'Background noise', 'Noise', 'Sound', 'Environmental noise',
  'Inside, small room', 'Inside, large room or hall',
  'Outside, urban or manmade', 'Outside, rural or natural',
  'Music', 'Speech', 'Narration, monologue',
]);

export class YamnetClassifier {
  constructor() {
    this.model = null;
    // Labels are bundled (no runtime network dependency) so a detection can
    // never come back as a meaningless "Sound 137".
    this.labels = YAMNET_LABELS;
  }

  async load(onProgress) {
    await tf.ready();
    onProgress?.('Loading AI model…');

    let lastErr;
    for (const { url, fromTFHub } of MODEL_URLS) {
      try {
        this.model = await tf.loadGraphModel(url, { fromTFHub });
        break;
      } catch (e) {
        lastErr = e;
        console.warn(`[YAMNet] Failed to load ${url}:`, e.message);
      }
    }
    if (!this.model) {
      throw new Error(`Could not load the sound model. ${lastErr?.message || ''}`.trim());
    }

    onProgress?.('AI ready');
  }

  predict(audioBuffer) {
    if (!this.model) return null;
    return tf.tidy(() => {
      const input = tf.tensor1d(audioBuffer).expandDims(0);
      const [scores] = this.model.predict(input);
      const meanScores = scores.mean(0);
      const scoresData = meanScores.dataSync();

      // Get top 5
      const top5 = Array.from(scoresData)
        .map((score, index) => ({ score, index }))
        .sort((a, b) => b.score - a.score)
        .slice(0, 5);

      // Pick highest-confidence specific label
      const best = top5.find(({ index }) => !GENERIC.has(this.labels[index] || '')) || top5[0];
      return {
        label: this.labels[best.index] || 'Sound',
        confidence: best.score,
        top5: top5.map(({ index, score }) => ({ label: this.labels[index], score })),
      };
    });
  }
}
